packages/misskey-js/src/acct.ts

```diff
@@ -3,13 +3,79 @@ export type Acct = {
 	host: string | null;
 };
 
-export function parse(_acct: string): Acct {
-	let acct = _acct;
-	if (acct.startsWith('@')) acct = acct.substring(1);
-	const split = acct.split('@', 2);
+export function correctAcct(acct: Acct, localHostname?: string): Acct {
+	const result = { ...acct };
+	if (!acct.host) {
+		result.host = null;
+	} else if (localHostname && acct.host === localHostname) {
+		result.host = null;
+	}
+	return result;
+}
+
+export function parse(acct: string): Acct {
+	let acctWithNoPrefix = acct;
+
+	if (acct.startsWith('@')) {
+		acctWithNoPrefix = acct.substring(1);
+	} else if (acct.startsWith('acct:')) {
+		acctWithNoPrefix = acct.substring(5);
+	}
+
+	const split = acctWithNoPrefix.split('@', 2);
+
 	return { username: split[0], host: split[1] || null };
 }
 
+export class UrlIsNotAcctLikeError extends Error {
+	constructor() {
+		super('This url is not acct like.');
+	}
+}
+
+/**
+ * Only supports `https?://example.com/@username` or `https?://example.com/@username@other.example.com`
+ */
+export function parseUrl(str: string): Acct {
+	const url = new URL(str);
+
+	if (url.hash.length > 0) throw new UrlIsNotAcctLikeError();
+	if (url.search.length > 0) throw new UrlIsNotAcctLikeError();
+
+	const splited = url.pathname.split('/');
+	if (splited.length !== 2) throw new UrlIsNotAcctLikeError();
+
+	const path = splited[1];
+
+	if (!path) throw new UrlIsNotAcctLikeError();
+	if (!path.startsWith('@')) throw new UrlIsNotAcctLikeError();
+	if (path.length <= 1) throw new UrlIsNotAcctLikeError();
+
+	const split = path.split('@', 3); // ['', 'username', 'other.example.com']
+
+	return { username: split[1], host: split[2] || url.hostname };
```

Comment:
```js
new URL("https://self.example.com/@username@ホゲ.jp")
```

This gives `@username@%E3%83%9B%E3%82%B2.jp` which is probably not what you want.